contracts/Issuehunter.sol

```diff
@@ -0,0 +1,291 @@
+    function Issuehunter() {
+        issueManager = msg.sender;
+        // The default reward period is one day
+        defaultRewardPeriod = 86400;
+        // The default execution period is one week.
+        defaultExecutePeriod = 604800;
+    }
+
+    /// Creates a new campaign.
+    function createCampaign(bytes32 issueId) {
+        // If the a campaigns for the selected issue exists already throws an
+        // exception.
+        require(campaigns[issueId].createdBy == 0);
+
+        campaigns[issueId] = Campaign({
+            executed: false,
+            total: 0,
+            createdBy: msg.sender,
+            rewardPeriodExpiresAt: 0,
+            executePeriodExpiresAt: 0,
+            resolutor: 0
+        });
+
+        CampaignCreated(issueId, msg.sender, now);
+    }
+
+    /// Add funds to the selected campaign.
+    //
+    // TODO: add issuehunter "tip". The tip could be used to compute the issue
+    // campaign's rank in the list. Higher tips will make campaigns more visible
+    // in the directory, by making them appear with a higher rank in the list.
+    function fund(bytes32 issueId) payable {
+        // Require that a campaign exists
+        require(campaigns[issueId].createdBy != 0);
+
+        // TODO: require that a campaign hasn't any verified resolution
+
+        // Add funds to the list, and update campaign's funds total amount
+        campaigns[issueId].funds[msg.sender] += msg.value;
+        campaigns[issueId].total += msg.value;
+
+        CampaignFunded(
+            issueId,
+            msg.sender,
+            now,
+            msg.value
+        );
+    }
+
+    // Submit a new resolution proposal.
+    //
+    // TODO: the submitter must pay a fee to let the issue manager send the
+    // transaction that verifies the proposed resolution.
+    function submitResolution(bytes32 issueId, bytes32 commitSHA) {
+        // Require that a campaign exists
+        require(campaigns[issueId].createdBy != 0);
+        // Fail if sender already submitted the same resolution
+        require(campaigns[issueId].resolutions[msg.sender] == 0 || campaigns[issueId].resolutions[msg.sender] != commitSHA);
+
+        // TODO: require that a campaign hasn't any verified resolution
+
+        campaigns[issueId].resolutions[msg.sender] = commitSHA;
+
+        ResolutionProposed(issueId, msg.sender, commitSHA);
+    }
+
+    // Verify a proposed resolution.
+    //
+    // Only the issue manager can invoke this function.
+    //
+    // The issue manager must verify that:
+    //
+    // 1. the proposed resolution is a real solution for the selected issue, for
+    //    instance by checking that the project's master branch includes the
+    //    commit SHA that has been proposed as a resolution
+    // 2. the resolution submitter address is included in one of the commit
+    //    messages in the branch that contains the resolution
+    function verifyResolution(bytes32 issueId, address resolutor) {
```

Comment:
Since a resolutor can submit more than one resolutions (and these will overwrite each other), I was thinking about race conditions here. Suppose a resolutor call the contract to submit a new resolution just before the issueManager call verifyResolution, what would you expect? Maybe we can also add the commitSHA here and fail if this is different from `campaigns[issueId].resolutions[resolutor]`

Reply:
I think each transaction is atomic and applied independently, but you're right that this function should take both an address and a commit sha as parameters. I will update it!

Reply:
@bugant I fixed `verifyResolution` to handle the special case that you suggested.